### PR TITLE
Multi-scale data loading on a "match" or "mix" schedule.

### PR DESCRIPTION
**Pull request by alxmrs**
A simpler alternative to #504. In our multi-scale model, we need one of two things: a `TrainData` batch that varies by scale per the sample index, or a "mixed" `TrainData` batch, where each Input/Label pair may come from one of the input scales. This new data loader accomplishes this need by providing a "schedule" interface and by delegating the actual loading to the existing Torch loader. In addition, this new load accomplishes a necessary IO-bound optimization: we only ever load the `TrainData`(s) batch that will actually get used. So, instead of loading `TrainData` batches for every scale per index, we only load the data source that actually gets returned for that index. For this "match" case, this means we only look up one data source, for the "mix" case, this means we look up two.

Last, since data loading is no longer one-to-one (a single index in a batch could refer to multiple sources of data), this PR also includes a new batch sampler that samples across equivalence classes.

### CODE REVIEW

File: src/ocean_emulators/constants.py [hunk 1]
```diff
@@ -234,6 +234,7 @@ def construct_metadata(data: xr.Dataset) -> dict[str, dict[str, str]]:
 
 class LoaderVersion(enum.Enum):
     OM4_TORCH = "om4-torch"
+    OM4_MULTI = "om4-multi"
```
Comment (alxmrs) [inline src/ocean_emulators/constants.py]: Question for the reviewer: should I create two loaders for multi (one per schedule)?

File: src/ocean_emulators/datasets.py [hunk 2]
```diff
@@ -421,7 +471,7 @@ def __init__(
         executor: ThreadPoolExecutor | None = None,
     ):
         super().__init__()
-        self.id = f"{self.__class__.__name__}_{str(id(self))}"
+        self.id = f"{self.__class__.__name__}({str(id(self))})"
```
Comment (alxmrs) [inline src/ocean_emulators/datasets.py:476]: Drive-by-refactor: I think this is nicer.

File: tests/test_datasets.py [hunk 3]
```diff
@@ -41,6 +46,16 @@ def inference_loader_pair(trainer_pair: TrainPair) -> tuple[TrainConfig, DataLoa
     return cfg, trainer.inference_loader
 
 
+def coarsen(
```
Comment (alxmrs) [inline tests/test_datasets.py]: TODO(alxmrs): Currently unused.

File: tests/test_datasets.py [hunk 4]
```diff
@@ -88,22 +103,45 @@ def make_loader(
                     )
                     for stride in cfg.data_stride
                 ]
-
-                data: ConcatDataset = ConcatDataset(dataset_list)
                 collate_fn = collate_raw_train_data
+            case LoaderVersion.OM4_MULTI:
+                srcs = [src, src]
```
Comment (alxmrs) [inline tests/test_datasets.py]: TODO(alxmrs): add coarsen here; it will mean tests need updating.

File: tests/test_datasets.py [hunk 5]
```diff
@@ -88,22 +103,45 @@ def make_loader(
...
+                        stride=stride,
+                    )
 
-                loader = TrainDataLoader(raw_loader, dataset_list, torch.device("cpu"))
-                yield loader
+                dataset_list = [
+                    MultiscaleTrainDataset(
+                        srcs=srcs,
+                        make_dataset=partial(make_dataset, stride=stride),
+                        schedule="match",
```
Comment (alxmrs) [inline tests/test_datasets.py]: TODO(alxmrs): we're only exercising one of two pathways.

File: src/ocean_emulators/utils/train.py [hunk 6]
```diff
@@ -36,6 +40,23 @@ def collate_raw_train_data(data: Sequence[RawTrainData]) -> RawTrainData:
     return batched_data
 
 
+def collate_raw_multiscale_train_data(
+    data: Sequence[RawMultiscaleTrainData],
+) -> RawMultiscaleTrainData:
+    batched_data = RawMultiscaleTrainData(data[0].dataset_id, {}, data[0].index)
```
Comment (alxmrs) [inline src/ocean_emulators/utils/train.py:50]: The index value here is suspect.

File: configs/test/train_default_2step.yaml [hunk 7]
```diff
@@ -7,7 +7,7 @@ disk_mode: true
 pin_mem: true
 save_freq: 5
 epochs: 1
-batch_size: 1
+batch_size: 2
```
Comment (alxmrs) [inline configs/test/train_default_2step.yaml]: Needed to test if this PR actually worked with realistic batch sizes.

File: src/ocean_emulators/utils/train.py [hunk 8]
```diff
@@ -36,6 +40,27 @@ def collate_raw_train_data(data: Sequence[RawTrainData]) -> RawTrainData:
     return batched_data
 
 
+def collate_raw_multiscale_train_data(
+    data: Sequence[RawMultiscaleTrainData],
+) -> RawMultiscaleTrainData:
+    # When using MultiscaleGroupedBatchSampler, items in a batch have the same multiplex
+    # position (idx % multiplex_size) but not necessarily the same exact index.
+    # For example, indices [0, 4, 8] all have multiplex position 0 when multiplex_size=4.
+    # We use data[0].index as the representative since all items map to the same position.
```
Comment (alxmrs) [inline src/ocean_emulators/utils/train.py]:
```suggestion
    # When using EquivalenceGroupedBatchSampler, items in a batch have the same group
    # position (idx % group_size) but not necessarily the same exact index.
    # For example, indices [0, 4, 8] all have group position 0 when group_size=4.
    # We use data[0].index as the representative since all items map to the same position.
```

File: tests/test_datasets.py [hunk 9]
```diff
@@ -251,32 +305,58 @@ def test_loader__data_shape(
...
+                }
+                observed_patterns = set(example_resolutions)
+                # All observed patterns must be valid
+                assert observed_patterns <= valid_patterns, (
+                    f"All resolutions must be valid members of the cartesian product for 'mix' schedule. "
+                    f"Valid patterns: {valid_patterns}, got {observed_patterns}, "
+                    f"invalid patterns: {observed_patterns - valid_patterns}"
+                )
+                # We should have at least 1 pattern
+                assert len(observed_patterns) > 0, "Should have at least one pattern"
```
Comment (alxmrs) [inline tests/test_datasets.py:359]: AI generated; I think this could use some cleaning up.

**Comment by alxmrs** [comment]
Closing in favor of #523.
